src/main/java/com/openlattice/chronicle/services/upload/AppDataUploadManagerImpl.java

```diff
@@ -0,0 +1,654 @@
+
+        Map<UUID, Set<Object>> entity = new HashMap<>();
+
+        entity.put( commonTasksManager.getPropertyTypeId( FULL_NAME_FQN ), ImmutableSet.of( appPackageName ) );
+        entity.put( commonTasksManager.getPropertyTypeId( TITLE_FQN ), ImmutableSet.of( appName ) );
+
+        return entity;
+    }
+
+    private EntityKey getDeviceEntityKey( UUID devicesESID, String deviceId ) {
+        return new EntityKey(
+                devicesESID,
+                deviceId
+        );
+    }
+
+    private EntityKey getParticipantEntityKey( UUID participantESID, String participantId ) {
+        return new EntityKey(
+                participantESID,
+                participantId
+        );
+    }
+
+    private Set<OffsetDateTime> getDateTimeValuesFromDeviceData( List<SetMultimap<UUID, Object>> data ) {
+        Set<OffsetDateTime> dateTimes = new HashSet<>();
+        data.forEach(
+                entity -> {
+                    // most date properties in the entity are of length 1
+                    for ( Object date : entity.get( commonTasksManager.getPropertyTypeId( DATE_LOGGED_FQN ) ) ) {
+                        OffsetDateTime parsedDateTime = OffsetDateTime
+                                .parse( date.toString() );
+
+                        // filter out problematic entities with dates in the sixties
+                        if ( parsedDateTime.isAfter( MINIMUM_DATE ) ) {
+                            dateTimes.add( parsedDateTime );
+                        }
+                    }
+                }
+        );
+        return dateTimes;
+    }
+
+    private Pair<Map<EntityKey, Map<UUID, Set<Object>>>, Set<Triple<EntityKey, EntityKey, EntityKey>>> getMetadataEntitiesAndEdges(
+            DataApi dataApi,
+            DataIntegrationApi integrationApi,
+            List<SetMultimap<UUID, Object>> data,
+            UUID organizationId,
+            UUID studyId,
+            UUID participantESID,
+            UUID participantEKID,
+            String participantId
+    ) {
+        // entity set ids
+        UUID metadataESID = commonTasksManager.getEntitySetId( organizationId, CHRONICLE, METADATA, METADATA_ES );
+        UUID hasESID = commonTasksManager.getEntitySetId( organizationId, CHRONICLE, HAS, HAS_ES );
+
+        checkNotNullUUIDs( ImmutableSet.of( metadataESID, hasESID ) );
+
+        Map<EntityKey, Map<UUID, Set<Object>>> entitiesByEntityKey = Maps.newHashMap();
+        Set<Triple<EntityKey, EntityKey, EntityKey>> edgesByEntityKey = Sets.newHashSet();
+
+        // get all dates in new data batch
+        Set<OffsetDateTime> pushedDateTimes = getDateTimeValuesFromDeviceData( data );
+        if ( pushedDateTimes.size() == 0 ) {
+            return null;
+        }
+
+        String firstDateTime = pushedDateTimes
+                .stream()
+                .min( OffsetDateTime::compareTo )
+                .orElse( null )
+                .toString();
```

Comment:
this is gonna NPE on null

Reply:
Since the set will never contain nulls, it's safe to remove this `.orElse`